server/src/main/java/org/elasticsearch/transport/InboundAggregator.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.CompositeBytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.lease.Releasables;
+
+import java.io.IOException;
+import java.util.ArrayList;
+
+public class InboundAggregator implements Releasable {
+
+    private ReleasableBytesReference firstContent;
+    private ArrayList<ReleasableBytesReference> contentAggregation;
+    private Header currentHeader;
+    private boolean isClosed = false;
+
+    public void headerReceived(Header header) {
+        ensureOpen();
+        if (currentHeader != null) {
+            currentHeader = null;
+            throw new IllegalStateException("Header already received.");
+        }
+
+        currentHeader = header;
+    }
+
+    public void aggregate(ReleasableBytesReference content) {
+        ensureOpen();
+        if (currentHeader == null) {
+            content.close();
+            throw new IllegalStateException("Received content without header");
```

Comment:
should this be an assertion instead, or when do we expect this to ever occur?

Reply:
Yeah I guess an assertion is fine since the decoder protects us from random bytes triggering this.